src/python/pants/backend/shell/shunit2_test_runner.py

```diff
@@ -70,24 +70,30 @@ class TestSetup:
     process: Process
 
 
-class Shunit2AlreadySourced(Exception):
+class ShellNotConfigured(Exception):
     pass
 
 
-class ShellNotConfigured(Exception):
-    pass
+SOURCE_SHUNIT2_REGEX = re.compile(rb"(?:source|\.)\s+[.${}/'\"\w]*shunit2\b['\"]?")
 
 
-def validate_source_shunit2_not_included(fc: FileContent, address: Address) -> None:
-    lines = fc.content.splitlines()
-    for i, line in enumerate(lines):
-        if re.search(rb"(?:source|\.) +[.${}/'\"\w]*shunit2\b", line):
-            raise Shunit2AlreadySourced(
-                f"The test file {fc.path} sources shunit2 on line {i + 1} with: "
-                f"{line.decode()}\n\n"
-                f"Please remove this line so that Pants can run shunit2 for the target {address}, "
-                "as Pants will download the shunit2 script and source it for you automatically."
-            )
+def add_source_shunit2(fc: FileContent) -> FileContent:
+    # NB: We always run tests from the build root, so we source `shunit2` relative to there.
+    source_line = b"source ./shunit2"
+
+    lines = []
+    already_had_source = False
+    for line in fc.content.splitlines():
+        if SOURCE_SHUNIT2_REGEX.search(line):
```

Comment:
This works, but running the re 2x isn't awesome. You could pass a function instead of the source_line literal though and have that function toggle a bool.

Reply:
We only do the regex twice for lines with `source ./shunit2` in it, which we expect to be 0-1 lines per file (or their file wouldn't have worked in the first place). Going to stick with this to Keep It Simple.